dhall/src/Dhall/Parser/Token.hs

```diff
@@ -525,10 +528,13 @@ decOctet =
       where
         predicate c = '\x30' <= c && c <= '\x35'
 
-regName :: Parser Text
-regName = star (satisfy predicate <|> pctEncoded)
+domain :: Parser Text
+domain = domainLabel <> star ("." <> domainLabel ) <> option "."
+
+domainLabel :: Parser Text
+domainLabel = alphaNum_ <> star (((alphaNum_ <|> plus "-") <> alphaNum_) <|> alphaNum_)
```

Comment:
Looking more closely at this, I believe the rule from the official grammar could be simplified.  I believe it should just be:

```abnf
domainlabel = 1*ALPHANUM *(1*"-" 1*ALPHANUM)
```

... which will also fix the remaining backtracking issues you're running into (and improve performance because it permits bulk parsing operators).